File: tests/mock_systemctl.py
# mock_systemctl.py
#
# Don't try to run this with sudo in background

import os, sys
import subprocess as SUBP

cmd_pipe_name = "/mock_systemctl_cmds.txt"
response_pipe_name = "/mock_systemctl_responses.txt"
msg_file = "/mock_systemctl_msgs.txt"

def remove_pipe (pipe_name):
  try:
    os.remove (pipe_name)
  except StandardError, e:
    pass

#remove_pipe (cmd_pipe_name)
#remove_pipe (response_pipe_name)

cp = None
rp = None

services = {}

test_services = []

def create_pipe (pipe_name):
	try:
	  os.mkfifo (pipe_name)
	except StandardError, e:
	  print "Error creating fifo", pipe_name
	  print e
	  sys.exit(4)

def open_pipe (pipe_name, mode):
	try:
	  return os.open (pipe_name, mode, 0666)
	except StandardError, e:
	  print "Error opening fifo", pipe_name
	  print e
	  return -1

#create_pipe (cmd_pipe_name)
#create_pipe (response_pipe_name)


def test_is_active(svc_name):
  if svc_name in test_services:
    call_name = svc_name + ".service"
    rtn = SUBP.call (["systemctl", "is-active", call_name])
    if rtn == 0:
      active = "1"
    else:
      active = "0"
    services[svc_name] = active
  else:
    active = services.get (svc_name, None)
  if active is None:
    os.write (rp, "0\n")
  else:
    os.write (rp, "%s\n" % active[0])

def start_service (svc_name):
  if svc_name in test_services:
    call_name = svc_name + ".service"
    SUBP.call (["systemctl", "start", call_name])
  services[svc_name] = "1"
  os.write (rp, "1\n")

def stop_service (svc_name):
  if svc_name in test_services:
    call_name = svc_name + ".service"
    SUBP.call (["systemctl", "stop", call_name])
  services[svc_name] = "0"
  os.write (rp, "0\n")

def invalid_command (line):
  print "Invalid command: ", line
  os.write (rp, "?\n")

home_dir = None

for arg in sys.argv[1:]:
  if not home_dir:
    home_dir = arg
    continue
  if arg.endswith (".service"):
    arg = arg[0:-8]
  test_services.append (arg)

if not home_dir:
  print "Expecting home dir as first cmd line argument"
  sys.exit

cmd_pipe_name = home_dir + cmd_pipe_name
response_pipe_name = home_dir + response_pipe_name
msg_file = home_dir + msg_file

msgf = open (msg_file, "w")
msgf.write ("Opening pipes\n")
msgf.close()

cp = open_pipe (cmd_pipe_name, os.O_RDONLY)
if cp < 0:
  sys.exit (4)
rp = open_pipe (response_pipe_name, os.O_WRONLY)
if rp < 0:
  os.close (cp)
  sys.exit (4)

msgf = open (msg_file, "w")
msgf.write ("Opened pipes\n")
msgf.close()

while (True):
  line = os.read (cp, 128)
  if not line:
    continue;
  cmd = line[0]
  svc_name = line[1:]
  pos = svc_name.find (".service")
  if (pos < 0):
    pos = svc_name.find ("\n")
  if pos < 0:
    invalid_command (line)
    continue
  svc_name = svc_name[0:pos]
  if cmd == "?":
    test_is_active (svc_name)
  elif cmd == "1":
    start_service (svc_name)
  elif cmd == "0":
    stop_service (svc_name)
  else:
    invalid_command (line)
#end while

File: tests/mock_systemctl.c
// mock_systemctl.c
//
// compile:
//   gcc -o mock_systemctl -lm mock_systemctl.c
//

#include <stdio.h>
#include <fcntl.h>
#include <string.h>
#include <stdbool.h>
#include <errno.h>
#include <stdlib.h>
#include <unistd.h>

#define NAME_BUFLEN 128
#define CMD_PIPE_NAME "mock_systemctl_cmds.txt"
#define RESPONSE_PIPE_NAME "mock_systemctl_responses.txt"

static char cmd_pipe_name[NAME_BUFLEN];
static char response_pipe_name[NAME_BUFLEN];

static char cmd_buf[NAME_BUFLEN];
static char response_buf[10];

static int cmd_pipe_fd = -1;
static int response_pipe_fd = -1;

static int open_pipe (const char *pipe_name, int mode)
{
	int pipe_fd = open (pipe_name, mode | O_SYNC, 0666);
	if (pipe_fd < 0) {
		printf ("Error(1) opening pipe %s\n", pipe_name);
	}
	return pipe_fd;
}

static int send_command (char cmd, const char *svc_name)
{
	int nbytes;

	cmd_pipe_fd = open_pipe (cmd_pipe_name, O_WRONLY);
	if (cmd_pipe_fd < 0)
		return 4;

	response_pipe_fd = open_pipe (response_pipe_name, O_RDONLY);
	if (response_pipe_fd < 0) {
		close (cmd_pipe_fd);
		return 4;
	}

	memset (cmd_buf, '\0', NAME_BUFLEN);
	cmd_buf[0] = cmd;
	sprintf (cmd_buf+1, "%s\n", svc_name);
	nbytes = write (cmd_pipe_fd, cmd_buf, NAME_BUFLEN);
	if (nbytes < 0) {
		printf ("Error writing to pipe %s\n", cmd_pipe_name);
		close (cmd_pipe_fd);
		close (response_pipe_fd);
		return 4;
	}
	nbytes = read (response_pipe_fd, response_buf, 2);
	close (cmd_pipe_fd);
	close (response_pipe_fd);
	if (nbytes < 0) {
		printf ("Error reading from pipe %s\n", response_pipe_name);
		return 4;
	}
	return 0;
}

int test_svc_active (const char *svc_name)
{
	int rtn = send_command ('?', svc_name);
	if (rtn != 0)
		return rtn;
	if (response_buf[0] == '1') // active
		return 0;
	return 1; // inactive
}

int main(int argc, char *argv[])
{
	const char *home_dir;
	const char *command;

	if (argc < 3) {
		printf ("Expecting 2 arguments:\n");
		printf (" 1) command\n");
		printf (" 2) service name\n");
		return EINVAL;
	}
	
	home_dir = getenv ("HOME");
	if (NULL == home_dir) {
		printf ("No $HOME defined\n");
		return ENOENT;
	}

	sprintf (cmd_pipe_name, "%s/%s", home_dir, CMD_PIPE_NAME);
	sprintf (response_pipe_name, "%s/%s", home_dir, RESPONSE_PIPE_NAME);

	command = argv[1];
	if (strcmp (command, "is-active") == 0)
		return test_svc_active (argv[2]);
	else if (strcmp (command, "start") == 0)
		return send_command ('1', argv[2]);
	else if (strcmp (command, "stop") == 0)
		return send_command ('0', argv[2]);
	else {
		printf ("Invalid command\n");
		return EINVAL;
	}
}
